## REGRESSION PROJECT

- Marjan Rezvani

In this project, first of all, I am going to do a web scraping that focused on real estate sales in different cities of The United States, as listed on the Century21 website(https://www.century21.com). I scraped the website using BeautifulSoup library.
Then, I will be analyzing a house price predication for finding out price of house on different parameters. I will do Exploratory Data Analysis, split the training and testing data, Model Evaluation and Predictions.



- Tools:
    - BeautifulSoup
    - Pandas
    - NumPy
    - Sklearn
    - matplotlib
    - Seaborn
    - statsmodel

## Web Scraping and Preprocessing Data

In [387]:
import requests
from bs4 import BeautifulSoup
import re
import pandas as pd

from sklearn.linear_model import LinearRegression 
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

import scipy.stats as scp
from   matplotlib  import pyplot   as plt
from   matplotlib  import patches  as pat

from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold, cross_val_score
from sklearn.metrics import mean_squared_error
import sklearn.linear_model as linear_model
from sklearn.preprocessing import StandardScaler
import scipy.stats as stats



plt.style.use('seaborn-whitegrid')

In [18]:
def extract_info_from_one_page(one_page):
    l = []
    for item in all:
        d = {}
        try:
            d['price'] = item.find("a",{"class":"listing-price"}).text.replace("\n", "").strip()
        except:
            d['price']= None
        try:
            d['beds'] = item.find("div",{"class":"property-beds"}).text.replace("\n", "").strip().split(" ")[0]
        except:
            d['beds']=None
        try:
            d['baths'] = item.find("div",{"class":"property-baths"}).text.replace("\n", "").strip().split(" ")[0]
        except:
            d['baths']=None
        try:
            d['size'] = item.find("div",{"class":"property-sqft"}).text.replace("\n", "").strip().split(" ")[0]
        except:
            d['size']=None
        try:
            d['address'] = item.find("div",{"class":"property-address"}).text.replace("\n", "").strip()
        except:
            d['address']=None
        try:
            d['city'] = item.find("div",{"class":"property-city"}).text.replace("\n", "").strip().split(" ")[0]
        except:
            d['city']=None
        try:
            d['state'] = item.find("div",{"class":"property-city"}).text.replace("\n", "").strip().split(" ")[1]
        except:
            d['state']=None
        try:
            d['zipcode']= item.find("div",{"class":"property-city"}).text.replace("\n", "").strip().split(" ")[2]
        except:
            d['zipcode']=None
        l.append(d) 
        
    return(pd.DataFrame(l))



In [441]:
###page = requests.get("https://www.century21.com/real-estate/stamford-ct/LCCTSTAMFORD/?ty=0")
#base_url = "https://www.century21.com/real-estate/stamford-ct/LCCTSTAMFORD/?ty="
#base_url = "https://www.century21.com/real-estate/jersey-city-nj/LCNJJERSEYCITY/?ty="

#base_url = "https://www.century21.com/real-estate/greenwich-ct/LCCTGREENWICH/?ty="
#base_url = "https://www.century21.com/real-estate/hamden-ct/LCCTHAMDEN/?ty="

#base_url = "https://www.century21.com/real-estate/austin-tx/LCTXAUSTIN/?ty="
#base_url = "https://www.century21.com/real-estate/dallas-tx/LCTXDALLAS/?ty="
#base_url = "https://www.century21.com/real-estate/houston-tx/LCTXHOUSTON/?ty="
#base_url = "https://www.century21.com/real-estate/riverside-ca/LCCARIVERSIDE/?ty="
#base_url = "https://www.century21.com/real-estate/chicago-il/LCILCHICAGO/?ty="
#base_url = "https://www.century21.com/real-estate/miami-fl/LCFLMIAMI/?ty="

#base_url = "https://www.century21.com/real-estate/philadelphia-pa/LCPAPHILADELPHIA/?ty="
#base_url = "https://www.century21.com/real-estate/charlotte-nc/LCNCCHARLOTTE/?ty="
#base_url = "https://www.century21.com/real-estate/nashville-tn/LCTNNASHVILLE/?ty="
#base_url = "https://www.century21.com/real-estate/atlanta-ga/LCGAATLANTA/?ty="
#base_url = "https://www.century21.com/real-estate/seattle-wa/LCWASEATTLE/?ty="
#base_url = "https://www.century21.com/real-estate/miami-fl/LCFLMIAMI/?ty="
#base_url = "https://www.century21.com/real-estate/orlando-fl/LCFLORLANDO/?ty="
#base_url = "https://www.century21.com/real-estate/bismarck-nd/LCNDBISMARCK/?ty="
#base_url = "https://www.century21.com/real-estate/helena-mt/LCMTHELENA/?ty="
#base_url = "https://www.century21.com/real-estate/fargo-nd/LCNDFARGO/?ty="
#base_url = "https://www.century21.com/real-estate/boston-ma/LCMABOSTON/?ty="
#base_url = "https://www.century21.com/real-estate/portland-or/LCORPORTLAND/?ty="
#base_url = "https://www.century21.com/real-estate/fresno-ca/LCCAFRESNO/?ty="
base_url = "https://www.century21.com/real-estate/sacramento-ca/LCCASACRAMENTO/?ty="

In [442]:
regex = re.compile('infinite-item property-card clearfix property-card-*')
pdf_all = pd.DataFrame()
page_num = 1
page = requests.get(base_url + str(page_num)) 
while('No Results Found' not in  str(page.content)):
    print(page_num)
    page = requests.get(base_url + str(page_num))
    soup=BeautifulSoup(page.content,"html.parser")
    all = soup.find_all("div", {"class" : regex})
    pdf = extract_info_from_one_page(all)
    pdf_all = pd.concat([pdf_all, pdf])
    page_num += 1
    

1
2
3
4
5
6


In [443]:
pdf_all

price beds baths   size                    address        city state  \
0     $958,540    5     3  2,968         3837 Watermist Way  Sacramento    CA   
1     $947,399    4     3  2,869     5643 Waterstone Street  Sacramento    CA   
2     $931,202    4     3  2,679     5649 Waterstone Street  Sacramento    CA   
3     $920,490    5     4  2,804         3843 Watermist Way  Sacramento    CA   
4     $908,142    5     3  3,425      5948 Fallstaff Street  Sacramento    CA   
..         ...  ...   ...    ...                        ...         ...   ...   
15    $215,000    4     2  1,512  3117  Federalist Lane #13  Sacramento    CA   
16  $1,295,000    6     4  3,989   7791 River Landing Drive  Sacramento    CA   
17  $1,150,000    3     3  2,286           2901 25th Street  Sacramento    CA   
18    $929,950    4     3  1,948           1357 48th Street  Sacramento    CA   
19    $899,000    4     3  2,442      408 Fairgrounds Drive  Sacramento    CA   

   zipcode  
0    95835  
1    95835  
2    95835  
3    95835  
4    95835  
..     ...  
15   95827  
16   95831  
17   95818  
18   95819  
19   95817  

[81 rows x 8 columns]

In [444]:
pdf_all.to_csv('Sacramento.csv')

In [446]:
Stamford = pd.read_csv('Stamford.csv')
Greenwich = pd.read_csv('Greenwich.csv')
Hamden = pd.read_csv('Hamden.csv')
Austin = pd.read_csv('Austin.csv')
Houston = pd.read_csv('Houston.csv')
Dallas = pd.read_csv('Dallas.csv')
Chicago = pd.read_csv('Chicago.csv')
Philadelphia = pd.read_csv('Philadelphia.csv')
Charlotte = pd.read_csv('Charlotte.csv')
Nashville = pd.read_csv('Nashville.csv')
Atlanta = pd.read_csv('Atlanta.csv')
Seattle = pd.read_csv('Seattle.csv')
Miami = pd.read_csv('Miami.csv')
riverside = pd.read_csv('riverside.csv')
Orlando = pd.read_csv('Orlando.csv')
Bismarck = pd.read_csv('Bismarck.csv')
Helena = pd.read_csv('Helena.csv')
Fargo = pd.read_csv('Fargo.csv')
Boston = pd.read_csv('Boston.csv')
Portland = pd.read_csv('Portland.csv')
Fresno = pd.read_csv('Fresno.csv')
Sacramento = pd.read_csv('Sacramento.csv')

In [481]:
data = pd.concat([Stamford, Greenwich, Hamden, Austin, Houston, Dallas, Chicago, Philadelphia, Charlotte, Nashville, Atlanta, Seattle, Miami, riverside, Orlando, Bismarck, Helena, Fargo, Boston, Portland, Fresno, Sacramento ], axis=0)

In [482]:
data = data.loc[:, ~data.columns.str.contains('^Unnamed')]

# EDA

In [483]:
data.head()

price  beds  baths   size                    address      city state  \
0  $1,299,000   4.0    3.0  2,688             28 Dunn Avenue  Stamford    CT   
1  $1,599,000   4.0    4.0  6,922     679 Hunting Ridge Road  Stamford    CT   
2    $265,000   2.0    1.0    910  87 Glenbrook Road Unit 5G  Stamford    CT   
3  $1,850,000   5.0    4.0  5,587       61 Nathan Hale Drive  Stamford    CT   
4    $629,900   3.0    2.0  2,620          76 Charles Street  Stamford    CT   

  zipcode  
0    6905  
1    6903  
2    6906  
3    6902  
4    6902

In [484]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1753 entries, 0 to 80
Data columns (total 8 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   price    1753 non-null   object 
 1   beds     1643 non-null   float64
 2   baths    1642 non-null   float64
 3   size     1509 non-null   object 
 4   address  1747 non-null   object 
 5   city     1753 non-null   object 
 6   state    1720 non-null   object 
 7   zipcode  1753 non-null   object 
dtypes: float64(2), object(6)
memory usage: 123.3+ KB


There are 1753 instances for dataset, and Total number of attributes equals 8, of which 2 is quantitative, 6 categorical.

In [485]:
df1 = data.dropna()

In [486]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1420 entries, 0 to 80
Data columns (total 8 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   price    1420 non-null   object 
 1   beds     1420 non-null   float64
 2   baths    1420 non-null   float64
 3   size     1420 non-null   object 
 4   address  1420 non-null   object 
 5   city     1420 non-null   object 
 6   state    1420 non-null   object 
 7   zipcode  1420 non-null   object 
dtypes: float64(2), object(6)
memory usage: 99.8+ KB


In [487]:
df1 = df1.loc[df1.price!='Ask', :]

In [488]:
df1.dtypes

price       object
beds       float64
baths      float64
size        object
address     object
city        object
state       object
zipcode     object
dtype: object

In [489]:
def clean_currency(x):
    """ If the value is a string, then remove currency symbol and delimiters
    otherwise, the value is numeric and can be converted
    """
    if isinstance(x, str):
        return(x.replace('$', '').replace(',', ''))
    return(x)

In [490]:
df1['price'] = df1['price'].apply(clean_currency).astype('float')


In [491]:
df1['beds'] = df1['beds'].astype('int')

In [492]:
df1.head()

price  beds  baths   size                    address      city state  \
0  1299000.0     4    3.0  2,688             28 Dunn Avenue  Stamford    CT   
1  1599000.0     4    4.0  6,922     679 Hunting Ridge Road  Stamford    CT   
2   265000.0     2    1.0    910  87 Glenbrook Road Unit 5G  Stamford    CT   
3  1850000.0     5    4.0  5,587       61 Nathan Hale Drive  Stamford    CT   
4   629900.0     3    2.0  2,620          76 Charles Street  Stamford    CT   

  zipcode  
0    6905  
1    6903  
2    6906  
3    6902  
4    6902

In [493]:
df1['baths'] = df1['baths'].astype('int')

In [494]:
df1['size'] = df1['size'].str.replace(',','').astype('int')

In [495]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1420 entries, 0 to 80
Data columns (total 8 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   price    1420 non-null   float64
 1   beds     1420 non-null   int64  
 2   baths    1420 non-null   int64  
 3   size     1420 non-null   int64  
 4   address  1420 non-null   object 
 5   city     1420 non-null   object 
 6   state    1420 non-null   object 
 7   zipcode  1420 non-null   object 
dtypes: float64(1), int64(3), object(4)
memory usage: 99.8+ KB


In [496]:
df1.describe()

price         beds        baths          size
count  1.420000e+03  1420.000000  1420.000000   1420.000000
mean   7.282098e+05     3.188028     2.242254   2197.767606
std    1.446101e+06     1.205109     1.042664   1725.144184
min    1.100000e+03     1.000000     1.000000    500.000000
25%    2.596750e+05     2.000000     2.000000   1314.000000
50%    4.990000e+05     3.000000     2.000000   1800.000000
75%    8.260000e+05     4.000000     3.000000   2532.000000
max    2.800000e+07     9.000000     9.000000  25200.000000

###  Adding some additional features

In [497]:
# Adding an additional feature, 'region'
df1['region']= 'east'

In [498]:
df1.loc[df1.state.isin(['TX', 'TN', 'FL', 'GA']), 'region']='south'

In [499]:
df1.loc[df1.state.isin(['IL', 'ND', 'MT']), 'region']='north'

In [500]:
df1.loc[df1.state.isin(['CA', 'WA', 'OR']), 'region']='west'

In [501]:
# Adding an additional feature, 'price/sqft'
pricepersqft = pd.Series(df1.price / df1.size)
df1['pricepersqft'] = pricepersqft

In [502]:
df1.columns

Index(['price', 'beds', 'baths', 'size', 'address', 'city', 'state', 'zipcode',
       'region', 'pricepersqft'],
      dtype='object')

In [508]:
df2 = df1.copy()

In [514]:
mask = df2['state'].isin(['TX', 'TN', 'FL', 'GA', 'IL', 'ND', 'MT', 'CA', 'WA', 'OR', 'CT', 'NC', 'MA', 'PA' ])
df3 = df2[mask]
df3.head()

price  beds  baths  size                    address      city state  \
0  1299000.0     4      3  2688             28 Dunn Avenue  Stamford    CT   
1  1599000.0     4      4  6922     679 Hunting Ridge Road  Stamford    CT   
2   265000.0     2      1   910  87 Glenbrook Road Unit 5G  Stamford    CT   
3  1850000.0     5      4  5587       61 Nathan Hale Drive  Stamford    CT   
4   629900.0     3      2  2620          76 Charles Street  Stamford    CT   

  zipcode region  pricepersqft  
0    6905   east    101.643192  
1    6903   east    125.117371  
2    6906   east     20.735524  
3    6902   east    144.757433  
4    6902   east     49.287950

In [515]:
df3.state.value_counts()

CA    223
TX    197
CT    187
FL    155
NC     94
MA     90
PA     76
IL     67
GA     63
WA     50
TN     37
ND     33
OR     28
MT     22
Name: state, dtype: int64

In [517]:
mask_2 = df3['city'].isin(['Stamford', 'Greenwich', 'Hamden', 'Austin', 'Houston', 'Dallas', 'Chicago', 'Philadelphia', 'Charlotte', 'Nashville', 'Atlanta', 'Seattle', 'Miami', 'riverside', 'Orlando', 'Bismarck', 'Helena', 'Fargo', 'Boston', 'Portland', 'Fresno', 'Sacramento' ])
df4 = df3[mask_2]
df4.head()

price  beds  baths  size                    address      city state  \
0  1299000.0     4      3  2688             28 Dunn Avenue  Stamford    CT   
1  1599000.0     4      4  6922     679 Hunting Ridge Road  Stamford    CT   
2   265000.0     2      1   910  87 Glenbrook Road Unit 5G  Stamford    CT   
3  1850000.0     5      4  5587       61 Nathan Hale Drive  Stamford    CT   
4   629900.0     3      2  2620          76 Charles Street  Stamford    CT   

  zipcode region  pricepersqft  
0    6905   east    101.643192  
1    6903   east    125.117371  
2    6906   east     20.735524  
3    6902   east    144.757433  
4    6902   east     49.287950

In [518]:
df4.city.value_counts()

Charlotte       94
Orlando         92
Austin          90
Boston          90
Stamford        89
Houston         88
Fresno          81
Philadelphia    76
Sacramento      69
Chicago         67
Atlanta         57
Hamden          53
Greenwich       45
Miami           45
Seattle         43
Nashville       37
Portland        26
Helena          22
Fargo           20
Bismarck        13
Dallas          13
Name: city, dtype: int64

In [519]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1210 entries, 0 to 80
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         1210 non-null   float64
 1   beds          1210 non-null   int64  
 2   baths         1210 non-null   int64  
 3   size          1210 non-null   int64  
 4   address       1210 non-null   object 
 5   city          1210 non-null   object 
 6   state         1210 non-null   object 
 7   zipcode       1210 non-null   object 
 8   region        1210 non-null   object 
 9   pricepersqft  1210 non-null   float64
dtypes: float64(2), int64(3), object(5)
memory usage: 104.0+ KB


In [531]:
df4 = df4.reset_index()

In [536]:
df4 = df4.drop(columns='index')

In [538]:
df4

price  beds  baths  size                    address        city  \
0     1299000.0     4      3  2688             28 Dunn Avenue    Stamford   
1     1599000.0     4      4  6922     679 Hunting Ridge Road    Stamford   
2      265000.0     2      1   910  87 Glenbrook Road Unit 5G    Stamford   
3     1850000.0     5      4  5587       61 Nathan Hale Drive    Stamford   
4      629900.0     3      2  2620          76 Charles Street    Stamford   
...         ...   ...    ...   ...                        ...         ...   
1205   215000.0     4      2  1512  3117  Federalist Lane #13  Sacramento   
1206  1295000.0     6      4  3989   7791 River Landing Drive  Sacramento   
1207  1150000.0     3      3  2286           2901 25th Street  Sacramento   
1208   929950.0     4      3  1948           1357 48th Street  Sacramento   
1209   899000.0     4      3  2442      408 Fairgrounds Drive  Sacramento   

     state zipcode region  pricepersqft  
0       CT    6905   east    101.643192  
1       CT    6903   east    125.117371  
2       CT    6906   east     20.735524  
3       CT    6902   east    144.757433  
4       CT    6902   east     49.287950  
...    ...     ...    ...           ...  
1205    CA   95827   west     16.823161  
1206    CA   95831   west    101.330203  
1207    CA   95818   west     89.984351  
1208    CA   95819   west     72.766041  
1209    CA   95817   west     70.344288  

[1210 rows x 10 columns]

In [543]:
# Saving the dataframe into a csv file

df4.to_csv("Output.csv")